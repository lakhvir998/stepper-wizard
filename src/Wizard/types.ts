import React from "react";

export interface StepProps {
  id: string | number;
  component: React.ReactNode | Function;
  label: string;
  index?: number;
  active?: boolean;
  nextHref?: string;
  prevHref?: string;
}

export interface WizardProps {
  steps: StepProps[];
  activeStep: number;
  showStepByStepLegend?: boolean;
  dark?: boolean;
  onStepChange: (step: number) => void;
  onSuccess?: () => void;
}

export interface StepperProps {
  index: number;
  activeStep: number;
}

export interface StyledStepProps {
  readonly isActive?: boolean;
  readonly isCompleted?: boolean;
}

export interface StyledButtonProps {
  readonly primary?: boolean;
}
