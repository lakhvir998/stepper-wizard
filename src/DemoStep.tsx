import React from "react";
import styled from "styled-components";

export interface DemoProps {
  title: string;
}

const Container = styled.div`
  display: flex;
`;

/**
 * Stepper Wizard
 */
const DemoStep: React.FC<DemoProps> = ({ title }) => {
  return (
    <Container>
      <h1>{title}</h1>
    </Container>
  );
};

export default DemoStep;
