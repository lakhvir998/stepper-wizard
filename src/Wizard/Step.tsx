import React from "react";
import styled from "styled-components";
import { StepProps } from "./types";

const Container = styled.div`
  display: flex;
  flex-direction: column;
`;

const Label = styled.h1``;

const StepContent = styled.div``;

/**
 * Stepp Component
 */
const Step: React.FC<StepProps> = ({ id, component, label, active }) => {
  if (!active) return null;
  return (
    <Container>
      <Label>{label}</Label>
      <StepContent>{component}</StepContent>
    </Container>
  );
};

export default Step;
