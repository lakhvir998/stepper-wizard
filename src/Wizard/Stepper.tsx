import React from "react";
import styled from "styled-components";
import { StepperProps, StyledStepProps } from "./types";

const Container = styled.div<StyledStepProps>`
  display: flex;
  position: relative;
  flex-grow: 1;
  flex-basis: 0;
  &:after {
    content: "";
    position: absolute;
    background: ${(props) =>
      props.isCompleted ? props.theme.primary : "#f1f1f1"};
    display: block;
    width: 100%;
    height: 5px;
    top: 10px;
    z-index: 1;
  }
  &:last-child {
    &::after {
      background: transparent;
    }
  }
`;

const Step = styled.span<StyledStepProps>`
  width: 20px;
  height: 20px;
  line-height: 22px;
  display: inline-block;
  border-radius: 25px;
  background: ${(props) => (props.isCompleted ? props.theme.primary : "#fff")};
  border: ${(props) =>
    `2px solid ${
      props.isActive ? props.theme.primary : props.theme.secondary
    }`};
  color: ${(props) =>
    props.isCompleted
      ? props.theme.primaryTextColor
      : props.isActive
      ? props.theme.primary
      : props.theme.secondary};
  z-index: 2;
  position: absolute;
  left: 0;
  text-align: center;
  font-size: 14px;
`;

/**
 * Stepper Wizard
 */
const Stepper: React.FC<StepperProps> = ({ index, activeStep }) => {
  return (
    <Container isCompleted={activeStep > index}>
      <Step isActive={activeStep === index} isCompleted={activeStep > index}>
        {index + 1}
      </Step>
    </Container>
  );
};

export default Stepper;
