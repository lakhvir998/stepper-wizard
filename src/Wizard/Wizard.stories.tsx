import React from "react";
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Meta, Story } from "@storybook/react/types-6-0";

import Wizard from "./Wizard";
import { WizardProps } from "./types";
import DemoStep from "../DemoStep";

export default {
  title: "Example/Wizard",
  component: Wizard,
} as Meta;

const Template: Story<WizardProps> = (args) => <Wizard {...args} />;

export const Light = Template.bind({});

Light.args = {
  steps: [
    {
      id: 1,
      component: <DemoStep title="step 1" />,
      label: "Step 1",
      index: 1,
    },
    {
      id: 2,
      component: <DemoStep title="step 2" />,
      label: "Step 2",
      index: 2,
    },
    {
      id: 3,
      component: <DemoStep title="step 3" />,
      label: "Step 3",
      index: 3,
    },
  ],
  activeStep: 0,
};

export const Dark = Template.bind({});

Dark.args = {
  steps: [
    {
      id: 1,
      component: () => <DemoStep title="step 1" />,
      label: "Step 1",
      index: 1,
    },
    {
      id: 2,
      component: () => <DemoStep title="step 2" />,
      label: "Step 2",
      index: 2,
    },
    {
      id: 3,
      component: () => <DemoStep title="step 3" />,
      label: "Step 3",
      index: 3,
    },
  ],
  activeStep: 0,
  dark: true,
};

export const WithLegend = Template.bind({});

WithLegend.args = {
  steps: [
    {
      id: 1,
      component: () => <DemoStep title="step 1" />,
      label: "Step 1",
      index: 1,
    },
    {
      id: 2,
      component: () => <DemoStep title="step 2" />,
      label: "Step 2",
      index: 2,
    },
    {
      id: 3,
      component: () => <DemoStep title="step 3" />,
      label: "Step 3",
      index: 3,
    },
  ],
  activeStep: 0,
  showStepByStepLegend: true,
};

export const CompleteStep = Template.bind({});

CompleteStep.args = {
  steps: [
    {
      id: 1,
      component: () => <DemoStep title="step 1" />,
      label: "Step 1",
      index: 1,
    },
    {
      id: 2,
      component: () => <DemoStep title="step 2" />,
      label: "Step 2",
      index: 2,
    },
    {
      id: 3,
      component: () => <DemoStep title="step 3" />,
      label: "Step 3",
      index: 3,
    },
  ],
  activeStep: 3,
  showStepByStepLegend: true,
};
