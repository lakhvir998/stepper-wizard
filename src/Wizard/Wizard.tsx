import React from "react";
import styled, { ThemeProvider } from "styled-components";
import { WizardProps, StyledButtonProps } from "./types";
import Step from "./Step";
import Stepper from "./Stepper";
import themes from "./theme";

const Container = styled.div`
  display: flex;
  flex-direction: column;
  padding: 15px;
  background-color: ${(props) => props.theme.backgroundColor};
  color: ${(props) => props.theme.primary};
  border: 1px solid #ddd;
`;

const StepperContainer = styled.div`
  display: flex;
  justify-content: space-between;
  margin-bottom: 20px;
`;

const ActionContainer = styled.div`
  display: flex;
  margin-top: 20px;
  justify-content: flex-end;
`;

const Button = styled.button<StyledButtonProps>`
  background: none;
  border: 1px solid
    ${(props) => (props.primary ? props.theme.primary : props.theme.secondary)};
  background-color: ${(props) =>
    props.primary ? props.theme.primary : props.theme.secondary};
  color: ${(props) => props.theme.primaryTextColor};
  padding: 10px 20px;
  font-size: 18px;
  margin-left: 10px;
`;

/**
 * Stepper Wizard
 */
const Wizard: React.FC<WizardProps> = ({
  steps,
  activeStep,
  dark = false,
  showStepByStepLegend = true,
  onStepChange,
  onSuccess = () => {},
}) => {
  const theme = dark ? themes.dark : themes.light;

  /**
   * Function move to next step
   * @return void
   */
  const handleNext = () => {
    const step = activeStep + 1;
    onStepChange(step);
    if (finalStep(step)) {
      onSuccess();
    }
  };

  /**
   * Function move to previous step
   * @return void
   */
  const handleBack = () => {
    onStepChange(activeStep - 1);
  };

  /**
   * Function to check if final step
   * @return boolean
   * @param {number} step
   */
  const finalStep = (step: number) => {
    return step >= steps.length;
  };

  return (
    <ThemeProvider theme={theme}>
      <Container>
        {showStepByStepLegend && (
          <StepperContainer>
            {steps.map((step, index) => (
              <Stepper
                key={`stepper-${index + 1}`}
                index={index}
                activeStep={activeStep}
              />
            ))}
          </StepperContainer>
        )}
        {steps.map((step, index) => (
          <Step {...step} key={`step-${index}`} active={activeStep === index} />
        ))}

        {finalStep(activeStep) && (
          <>
            <h1>All Done!</h1>
            <p> You have successfully completed all steps.</p>
          </>
        )}
        {steps.length > activeStep && (
          <ActionContainer>
            {activeStep >= 1 && (
              <Button onClick={handleBack} data-testid="previous">
                Back
              </Button>
            )}
            <Button primary onClick={handleNext} data-testid="next">
              Next
            </Button>
          </ActionContainer>
        )}
      </Container>
    </ThemeProvider>
  );
};

export default Wizard;
