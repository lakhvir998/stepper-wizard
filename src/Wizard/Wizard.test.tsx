import React from "react";
import { shallow } from "enzyme";
import { render, fireEvent } from "@testing-library/react";
import Wizard from "./Wizard";
import { WizardProps } from "./types";
import DemoStep from "../DemoStep";

const steps = [
  {
    id: 1,
    component: () => <DemoStep title="step 1" />,
    label: "Step 1",
    index: 1,
  },
  {
    id: 2,
    component: () => <DemoStep title="step 2" />,
    label: "Step 2",
    index: 2,
  },
  {
    id: 3,
    component: () => <DemoStep title="step 3" />,
    label: "Step 3",
    index: 3,
  },
];

const renderWizard = (props: Partial<WizardProps> = {}) => {
  const defaultProps: WizardProps = {
    steps,
    activeStep: 0,
    onStepChange() {
      return;
    },
  };
  return render(<Wizard {...defaultProps} {...props} />);
};

describe("<Wizard />", () => {
  test("renders the component", () => {
    const component = shallow(
      <Wizard
        steps={steps}
        activeStep={0}
        showStepByStepLegend
        onStepChange={(step) => console.log(step)}
        onSuccess={() => console.log("success")}
      />
    );
    expect(component).toMatchSnapshot();
  });

  test("should change next step", async () => {
    const onStepChange = jest.fn();
    const { findByTestId } = renderWizard({
      onStepChange,
    });
    const next = await findByTestId("next");

    fireEvent.click(next);

    expect(onStepChange).toHaveBeenCalledWith(1);
  });

  test("should change previous step", async () => {
    const onStepChange = jest.fn();
    const { findByTestId } = renderWizard({
      onStepChange,
      activeStep: 1,
    });
    const next = await findByTestId("previous");

    fireEvent.click(next);

    expect(onStepChange).toHaveBeenCalledWith(0);
  });
});
