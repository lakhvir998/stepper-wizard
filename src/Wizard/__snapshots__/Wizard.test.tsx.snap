// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Wizard /> renders the component 1`] = `
<Component
  theme={
    Object {
      "backgroundColor": "#ffffff",
      "primary": "#302AE6",
      "primaryTextColor": "#ffffff",
      "secondary": "#536390",
    }
  }
>
  <styled.div>
    <styled.div>
      <Stepper
        activeStep={0}
        index={0}
        key="stepper-1"
      />
      <Stepper
        activeStep={0}
        index={1}
        key="stepper-2"
      />
      <Stepper
        activeStep={0}
        index={2}
        key="stepper-3"
      />
    </styled.div>
    <Step
      active={true}
      component={[Function]}
      id={1}
      index={1}
      key="step-0"
      label="Step 1"
    />
    <Step
      active={false}
      component={[Function]}
      id={2}
      index={2}
      key="step-1"
      label="Step 2"
    />
    <Step
      active={false}
      component={[Function]}
      id={3}
      index={3}
      key="step-2"
      label="Step 3"
    />
    <styled.div>
      <styled.button
        data-testid="next"
        onClick={[Function]}
        primary={true}
      >
        Next
      </styled.button>
    </styled.div>
  </styled.div>
</Component>
`;

exports[`renders the component 1`] = `
<Component
  theme={
    Object {
      "backgroundColor": "#ffffff",
      "primary": "#302AE6",
      "primaryTextColor": "#ffffff",
      "secondary": "#536390",
    }
  }
>
  <styled.div>
    <styled.div>
      <Stepper
        activeStep={0}
        index={0}
        key="stepper-1"
      />
      <Stepper
        activeStep={0}
        index={1}
        key="stepper-2"
      />
      <Stepper
        activeStep={0}
        index={2}
        key="stepper-3"
      />
    </styled.div>
    <Step
      active={true}
      component={[Function]}
      id={1}
      index={1}
      key="step-0"
      label="Step 1"
    />
    <Step
      active={false}
      component={[Function]}
      id={2}
      index={2}
      key="step-1"
      label="Step 2"
    />
    <Step
      active={false}
      component={[Function]}
      id={3}
      index={3}
      key="step-2"
      label="Step 3"
    />
    <styled.div>
      <styled.button
        onClick={[Function]}
        primary={true}
      >
        Next
      </styled.button>
    </styled.div>
  </styled.div>
</Component>
`;
